### ISSUE TITLE
Detection quality is not too great

### ISSUE THREAD

**Issue by karanchahal** [facebookresearch/silk#17]
The SILK keypoints that I see are good at first glance but when I try to use it for feature keypoint matching, they dont seem very good. THere are small errors that mess up the 3d reconstruction. Any suggestions to improve this ?

**Comment by gleize** [comment]
Hi @karanchahal,

* Is the error a matching error ? Or a keypoint position error (i.e. shifted from where they should be) ?
* What setup are you using ? (model, checkpoint, image size, top-k, threshold, matching)
* Could you send the part of the code where you run SiLK ? It's difficult to pinpoint the issue without too much info.

In general, there are several factors that can be tuned for better performance :
* `Image size`. Too small and positions won't be accurate enough. Too large and the top-k keypoints from both image are less likely to overlap. There is usually a sweet spot.
* `Top-k`. Too small and the keypoints won't be repeatable. Too large and it will cost more memory and computation.
* `Matching`. Using MNN bare, or with ratio-test or double-softmax will also affect performance. On IMC, we used a double-softmax with threshold of 0.99, and temperature of 0.1. This might help in your case.
* `RANSAC/MAGSAC` parameters. Keypoints methods tend to be quite sensitive to those parameters (mostly the inlier threshold and max iterations).

**Comment by karanchahal** [comment]
Sure I can paste my code and result if that helps:

```silk_detector.py

# Copyright (c) Meta Platforms, Inc. and affiliates.
# All rights reserved.

# This source code is licensed under the license found in the
# LICENSE file in the root directory of this source tree.

import os
from copy import deepcopy

import numpy as np
import skimage.io as io
import torch

from silk.backbones.silk.silk import from_feature_coords_to_image_coords
from silk.cli.image_pair_visualization import create_img_pair_visual, save_image, img_pair_visual

from silk.backbones.silk.silk import SiLKVGG as SiLK
from silk.backbones.superpoint.vgg import ParametricVGG

from silk.config.model import load_model_from_checkpoint
from silk.models.silk import matcher

import os
from keypoint_saver import KeypointHandler

class SilkConfig:
    def __init__(self):
        self.checkpoint_path = [silk/assets/models/silk/analysis/alpha/pvgg-4.ckpt"](url)
        self.device = "cpu"
        self.nms = 0
        self.border = 0
        self.threshold = 1.0
        self.top_k = 100
        self.default_output = (
            "sparse_positions",
            "sparse_descriptors",
        )
        self.scale_factor = 1.41
        self.backbone = ParametricVGG(
            use_max_pooling=False,
            padding=0,
            normalization_fn=[torch.nn.BatchNorm2d(i) for i in (64, 64, 128, 128)],
        )
        self.matcher = matcher(postprocessing="ratio-test", threshold=0.6)

def load_images(*paths, config, as_gray=True):
    images = np.stack([io.imread(path, as_gray=as_gray) for path in paths])
    images = torch.tensor(images, device=config.device, dtype=torch.float32)
    if not as_gray:
        images = images.permute(0, 3, 1, 2)
        images = images / 255.0
    else:
        images = images.unsqueeze(1)  # add channel dimension
    return images

def get_model(config):
    # Load model
    model = SiLK(
        in_channels=1,
        backbone=deepcopy(config.backbone),
        detection_threshold=config.threshold,
        detection_top_k=config.top_k,
        nms_dist=config.nms,
        border_dist=config.border,
        default_outputs=config.default_output,
        descriptor_scale_factor=config.scale_factor,
        padding=0,
    )
    model = load_model_from_checkpoint(
        model,
        checkpoint_path=config.checkpoint_path,
        state_dict_fn=lambda x: {k[len("_mods.model.") :]: v for k, v in x.items()},
        device=config.device,
        freeze=True,
        eval=True,
    )
    return model
```

the code that generates the matches:

```

def handle_silk(args, current_run_dir):
    """
    Computes SILK keypoints and saves matched image in the folder
    """
    image_1_path, image_2_path = args.image1, args.image2
    config = SilkConfig()
    # load image
    images_0 = load_images(image_1_path, config=config)
    images_1 = load_images(image_2_path, config=config)

    # load model
    model = get_model(config)

    # run model
    sparse_positions_0, sparse_descriptors_0 = model(images_0)
    sparse_positions_1, sparse_descriptors_1 = model(images_1)

    sparse_positions_0 = from_feature_coords_to_image_coords(
        model, sparse_positions_0)
    sparse_positions_1 = from_feature_coords_to_image_coords(
        model, sparse_positions_1)

    # get matches
    matches = config.matcher(sparse_descriptors_0[0], sparse_descriptors_1[0])

    # create output image
    image_pair = create_img_pair_visual(
        image_1_path,
        image_2_path,
        None,
        None,
        sparse_positions_0[0][matches[:, 0]].detach().cpu().numpy(),
        sparse_positions_1[0][matches[:, 1]].detach().cpu().numpy(),
    )
    # Save image for matches
    save_image(
        image_pair,
        current_run_dir,
        "silk_matches.png",
    )

    return sparse_positions_0, sparse_positions_1, sparse_descriptors_0, sparse_descriptors_1

```

**Comment by karanchahal** [comment]
Is the error a matching error ? Or a keypoint position error (i.e. shifted from where they should be) ?
- I think it might be a matching error, because the keypoints seem to be shifted slightly.
What setup are you using ? (model, checkpoint, image size, top-k, threshold, matching)
- See code
Could you send the part of the code where you run SiLK ? It's difficult to pinpoint the issue without too much info.
- Done.
THe issues could be top k, maybe, I am using sparse keypoints becuase I can't debug dense matches visually. So I am building a feature tracker on top of this and after 5-6 frames, the keypoints fail to match across all frames even though they should.

**Comment by gleize** [comment]
Thanks @karanchahal for providing the code.

The top-k is definitely too low in your case. SiLK tends to provide many good keypoints.
For example, if you have two images, and each have 5k keypoints with a score around 0.9, there is no guarantee that the keypoint ranking will be conserved between those two images. Therefore selecting only 100 (2%) keypoints out of 5k will likely result in little overlapping between the two sets. Causing erroneous matching.

If the main reason for this choice of small top-k is debugging purpose, you might want to increase the top-k to at least 5k, and then only display the best 100 keypoints after matching / inlier selection.

Another suggestion would be to tune the ratio-test threshold. Low value will get you better matching precision, but high value will get you better pose estimation (if you're doing 3D reconstruction, I'm guessing you're estimating pose as well as depth).

**Comment by gleize** [comment]
Answers added to FAQ. Closing now.
